File: java/src/medium/Partition.java
package medium;

import java.util.ArrayList;
import java.util.List;

public class Partition {
    private boolean isPalindrome(String s, int left, int right) {
        while (left < right) {
            if (s.charAt(left) != s.charAt(right)) {
                return false;
            }
            left++;
            right--;
        }
        return true;
    }

    private enum Status {
        UNKNOWN,
        YES,
        NO,
    }

    private Status[][] dp;
    private List<List<String>> res = new ArrayList<>();

    private void backTrace(String s, int start, List<String> cur) {
        if (start == s.length()) {
            res.add(new ArrayList<>(cur));
            return;
        }
        for (int i = start; i < s.length(); i++) {
            if (dp[start][i] == Status.UNKNOWN) {
                boolean temp = isPalindrome(s, start, i);
                if (temp) {
                    dp[start][i] = Status.YES;
                } else {
                    dp[start][i] = Status.NO;
                }
            }
            if (dp[start][i] == Status.YES) {
//                cur.addLast(s.substring(start, i + 1));
                backTrace(s, i + 1, cur);
//                cur.removeLast();

            }
        }
    }

    public List<List<String>> partition(String s) {
        dp = new Status[s.length()][s.length()];
        for(int i = 0; i < s.length(); i++) {
            for(int j = 0; j < s.length(); j++) {
                dp[i][j] = Status.UNKNOWN;
            }
        }
        backTrace(s, 0, new ArrayList<>(s.length()));
        return res;
    }

    public static void main(String[] args) {
        System.out.println(new Partition().partition("aab"));
    }
}

File: typescript/medium/partition.ts
function isPart(s: string, left: number, right: number): boolean {
  while (left < right) {
    if (s[left] != s[right]) {
      return false;
    }
    left++;
    right--;
  }
  return true;
}

let dp: number[][];
let cur: string[];
let res: string[][];
function partition(s: string): string[][] {
  dp = new Array<Array<number>>(s.length);
  res = new Array();
  for (let i = 0; i < dp.length; i++) {
    dp[i] = new Array<number>(s.length);
    dp[i].fill(-1);
  }
  cur = new Array<string>();
  dfs(s, 0);
  return res;
}

function dfs(str: string, start: number) {
  if (start == str.length) {
    const add = new Array(cur.length);
    for (let i = 0; i < add.length; i++) {
      add[i] = cur[i];
    }
    res.push(cur);
    return;
  }

  for (let i = start; i < str.length; i++) {
    if (dp[start][i] == -1) {
      const r = isPart(str, start, i);
      if (r) {
        dp[start][i] = 1;
      } else {
        dp[start][i] = 0;
      }
    }
    if (dp[start][i] == 1) {
      cur.push(str.substring(start, i + 1));
      dfs(str, i + 1);
      cur.pop();
    }
  }
}
